cmd/limactl/prune.go

```diff
@@ -16,15 +22,97 @@ func newPruneCommand() *cobra.Command {
 		RunE:              pruneAction,
 		ValidArgsFunction: cobra.NoFileCompletions,
 	}
+
+	pruneCommand.Flags().BoolP("fallback", "F", false, "Only prune images without a digest specified (fallback images usually)")
+	pruneCommand.Flags().BoolP("unreferenced", "U", false, "Only prune downloads not referenced in any VM or template")
 	return pruneCommand
 }
 
 func pruneAction(cmd *cobra.Command, args []string) error {
+	pruneFallback, err := cmd.Flags().GetBool("fallback")
+	if err != nil {
+		return err
+	}
+	pruneUnreferenced, err := cmd.Flags().GetBool("unreferenced")
+	if err != nil {
+		return err
+	}
+
+	if pruneFallback || pruneUnreferenced {
+		files, err := getReferencedDownloads()
+		if err != nil {
+			return err
+		}
+
+		cacheEntries, err := downloader.CachedDownloads(downloader.WithCache())
+		if err != nil {
+			return err
+		}
+
+		logrus.Debug("Cache entries:")
+		for _, entry := range cacheEntries {
+			logrus.Debugf("+ %q => %q", entry.ID, entry)
+
+			// check if the cache entry is referenced
+			if refFile, refFound := files[entry.ID]; refFound {
+				if refFile.Location != entry.Location { // sanity check
+					logrus.Warnf("Ref'd URL %q != cached URL %q for %q.", entry.ID, refFile.Location, entry.Location)
+				}
+
+				if pruneFallback && refFile.Digest == "" {
+					// delete the fallback image entry (entry w/o digest) even if referenced
+					logrus.Infof("Deleting fallback entry %q (%q)", entry.Path, entry.Location)
+					if err := os.RemoveAll(entry.Path); err != nil {
+						logrus.WithError(err).Errorf("Cannot delete directory %q. Skipping...", entry.Path)
+					}
+				}
+			} else {
+				if pruneUnreferenced {
+					// delete the unreferenced cached entry
+					logrus.Infof("Deleting unreferenced entry %q (%q)", entry.Path, entry.Location)
+					if err := os.RemoveAll(entry.Path); err != nil {
+						logrus.WithError(err).Errorf("Cannot delete directory %q. Skipping...", entry.Path)
+					}
+				}
+			}
+		}
+		logrus.Debug(".")
+		return nil
+	}
+
+	// prune everything if no options specified
 	ucd, err := os.UserCacheDir()
 	if err != nil {
 		return err
 	}
 	cacheDir := filepath.Join(ucd, "lima")
-	logrus.Infof("Pruning %q", cacheDir)
+	logrus.Infof("Pruning everything in %q", cacheDir)
 	return os.RemoveAll(cacheDir)
 }
+
+func makeDigest(s string) string {
+	return fmt.Sprintf("%x", sha256.Sum256([]byte(s)))
+}
```

Comment:
go-digest might be better

Reply:
I agree.. but then this line should be changed too: https://github.com/lima-vm/lima/blob/master/pkg/downloader/downloader.go#L167

Reply:
Not sure about this one: https://github.com/lima-vm/lima/blob/master/pkg/limayaml/defaults.go#L53

Reply:
I dropped the `makeDigest()` function as it was used just in single place after the refactoring. Also `sha256.sha256()` is replaced with the `digest` module functions chain. Though I'm not sure if downloader.go:167 should be replaced too.

Reply:
@AkihiroSuda please let me know if I should update the https://github.com/lima-vm/lima/blob/master/pkg/downloader/downloader.go#L180 to use `digest` too.. I'm a bit worrying about inconsistency possible..

Reply:
New code is using `digest` instead of sprintf()'ed sha256. Old code in downloader.go still uses the sha256 call.